src/click/parser.py

```diff
@@ -337,11 +337,27 @@ def _match_long_opt(self, opt, explicit_value, state):
             if explicit_value is not None:
                 state.rargs.insert(0, explicit_value)
 
+            # if the option is a prompt and `prompt_required` is false,
+            # then the current rarg will serve as a flag, meaning the
+            # user will be prompted for an input if no argument is passed.
             nargs = option.nargs
             if len(state.rargs) < nargs:
-                _error_opt_args(nargs, opt)
+                if option.obj.prompt is not None and not option.obj.prompt_required:
+                    value = option.obj.prompt_for_value(self.ctx)
+                else:
+                    _error_opt_args(nargs, opt)
             elif nargs == 1:
-                value = state.rargs.pop(0)
+                next_rarg = state.rargs[0]
```

Comment:
I've noticed in the current codebase that this code block in `_match_long_opts` is identical in `_match_short_opts`. I kept it this way when making changes and just added the changes in both places. Should this be factored out, or is it not that significant? 

https://github.com/pallets/click/blob/637ea206c348100582beda3c76a8c51e9d7437ab/src/click/parser.py#L341-L347